### PR TITLE
Only allocate CO dockets once per week

### PR DESCRIPTION
**Pull request by rubaiyat22**
Resolves https://vajira.max.gov/browse/CASEFLOW-252

### Description
Previously, we'd allocate at most 4 docket days for CO hearing per week. This PR changes this to only create at most 1 docket day for CO hearing per week. We and the branch chose Wednesday for the day unless wednesday is a holiday, weekend or not available in which case we pick another non-holiday. 

- write new test
- update old test expectations

### Testing Plan
1. Login as `BVASYELLOW` and visit http://localhost:3000/hearings/schedule/build/upload
2. Upload `blankRoSpreadsheet.xlsx` for `RO/CO hearings`; NOTE: you can find `blankRoSpreadsheet.xlsx` within caseflow dir;
   You could also upload `validRoSpreadsheet.xlxs` but start and end dates would have to be within 1/1/18 - 6/1/18
3. Click `Continue` and ensure that correct number of CO docket days would be created for the days you've provided

### CODE REVIEW

File: spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb [hunk 1]
```diff
@@ -265,5 +265,13 @@
 
       it { expect { subject }.to raise_error(HearingSchedule::GenerateHearingDaysSchedule::NoDaysAvailableForRO) }
     end
+
+    context "allocated days to Central Office" do
+      subject { generate_hearing_days_schedule.generate_co_hearing_days_schedule }
+
+      it "only allocates 1 docket per week" do
+        expect(subject.count).to eq(25) # 26 wednesdays between 2018-04-01 and 2018-09-30; 1 holiday on Wed, July 4
```
Comment (ferristseng) [inline spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb:273]: Should we pick a different CO day if the Wednesday falls on a holiday for the week?

Comment (rubaiyat22) [reply spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb:273]: i think we should!

Comment (rubaiyat22) [reply spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb:273]: this will require a bit more work but i can do that and tag you back

Comment (ferristseng) [reply spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb:273]: excellent, thank you! everything else looks great!

Comment (rubaiyat22) [reply spec/services/hearing_schedule/generate_hearing_days_schedule_spec.rb]: made those changes

File: app/services/hearing_schedule/generate_hearing_days_schedule.rb [hunk 2]
```diff
@@ -93,15 +94,25 @@ def sort_ros_by_rooms_and_allocated_days
...
+    co_schedule_args = {
+      request_type: HearingDay::REQUEST_TYPES[:central],
+      room: "2",
+      bva_poc: "CAROL COLEMAN-DEW"
+    }
+
     (@schedule_period.start_date..@schedule_period.end_date).each do |scheduled_for|
-      next unless valid_day_to_schedule_co(scheduled_for)
+      # if CO_DAYS_OF_WEEK falls on a holiday, pick a day of the week that isn't a holiday
+      if CO_DAYS_OF_WEEK.include?(scheduled_for.cwday) && holiday?(scheduled_for)
```
Comment (ferristseng) [inline app/services/hearing_schedule/generate_hearing_days_schedule.rb:105]: I think this should also check if `co_not_available?`. We could also do a sanity check here for `weekend?` but that might not be totally necessary.

Comment (ferristseng) [reply app/services/hearing_schedule/generate_hearing_days_schedule.rb:105]: There might be an edge case here that this doesn't handle if the end of the schedule period ends on a date that is before Wednesday (Monday or Tuesday). I'm not sure if this would ever come up in practice, but it might be good to still code for.

This also could be a non-issue because the Wednesday would appear in the subsequent schedule period...but then the date arithmetic logic in `get_fallback_date_for_co` might need to be bounded to the schedule period.

File: spec/controllers/hearings/schedule_periods_controller_spec.rb [hunk 3]
```diff
@@ -117,7 +117,7 @@
       @controller = Hearings::HearingDayController.new
       get :index, params: { start_date: "2018-01-01", end_date: "2018-06-01" }, as: :json
       expect(response).to be_successful
-      expect(JSON.parse(response.body)["hearings"].size).to eq(427)
+      expect(JSON.parse(response.body)["hearings"].size).to eq(365)
```
Comment (ferristseng) [inline spec/controllers/hearings/schedule_periods_controller_spec.rb:120]: I was trying to calculate why this went from 427 => 365. I think this should be 364?

21 weeks between 2018-01-01 and 2018-06-01
3 fewer days per week
21 * 3 = 63
427 - 63 = 364

I assumed the 365 was due to a Wednesday holiday before, but it looks like the tests are still passing

**Review by ferristseng — CHANGES_REQUESTED** [review]
I think I found an edge case and noted some other small tweaks.

Regrading the edge case, I noticed in one of the tests, the schedule period ends on a Thursday, so I'm wondering if it's possible to have a schedule period end on any day of the week? If so, I think the edge case I mention is worth raising to St*ve, and getting input on from the Board.

File: spec/controllers/hearings/schedule_periods_controller_spec.rb (hunk 3 above)
Comment (rubaiyat22) [reply spec/controllers/hearings/schedule_periods_controller_spec.rb:120]: here's the calculation that i did which returned 364
previously before this change there were 68 invalid dates (meaning `valid_day_to_schedule_co(scheduled_for)` returned false; collected by running the loop)
there are 152 weekdays in total between 2018-01-01 and 2018-06-01 (https://www.timeanddate.com/date/workdays.html?d1=1&m1=1&y1=2018&d2=1&m2=6&y2=2018&ti=on&) 
152 - 68 = 84 
there are 22 wednesdays between 2018-01-01 and 2018-06-01
84 - 22 = 62 
62 is the total number of difference between prev and current version
let me know if you'd like to look at this together

File: app/services/hearing_schedule/generate_hearing_days_schedule.rb (hunk 2 above)
Comment (rubaiyat22) [reply app/services/hearing_schedule/generate_hearing_days_schedule.rb]: Spoke to Steve and we decided this is fine as is for now

File: spec/controllers/hearings/schedule_periods_controller_spec.rb (hunk 3 above)
Comment (ferristseng) [reply spec/controllers/hearings/schedule_periods_controller_spec.rb:120]: ok, I think this math checks out. I think I was not accounting for 22 wednesdays!

File: app/services/hearing_schedule/generate_hearing_days_schedule.rb [hunk 4]
```diff
@@ -331,11 +341,23 @@ def co_not_available?(day)
...
 
+  def weekend_or_holiday_or_not_available?(date)
+    weekend?(date) || holiday?(date) || co_not_available?(date)
+  end
+
+  # pick the first day from fallback days that is valid
+  def get_fallback_date_for_co(scheduled_for)
+    valid_cwday = CO_FALLBACK_DAYS_OF_WEEK.detect do |cwday|
+      date = scheduled_for.beginning_of_week + (cwday - 1).day
+
```
Comment (ferristseng) [inline app/services/hearing_schedule/generate_hearing_days_schedule.rb:352]: I think there still needs to be a check here to ensure that fallback date falls within the schedule period. The example is if the schedule period starts on a Thursday, the fallback date could be computed as Monday, due to the call to `beginning_of_week`

**Review by ferristseng — APPROVED** [review]
🚢 LGTM!
